### PR TITLE
SONAR-6571 WS api/metrics/create create or reactivate a custom metric

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+  public void handle(Request request, Response response) throws Exception {
+    userSession.checkLoggedIn().checkGlobalPermission(GlobalPermissions.SYSTEM_ADMIN);
+    String key = request.mandatoryParam(PARAM_KEY);
+
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      MetricDto metricInDb = dbClient.metricDao().selectNullableByKey(dbSession, key);
+      checkMetricInDb(metricInDb, key);
+
+      MetricDto metricTemplate = new MetricDto()
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java]: CRITICAL Possible null pointer dereference in org.sonar.server.metric.ws.CreateAction.handle(Request, Response) due to return value of called method [rule](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 new issue:
* 1 critical

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+public class CreateAction implements MetricsWsAction {
+  private static final String ACTION = "create";
+  public static final String PARAM_NAME = "name";
+  public static final String PARAM_KEY = "key";
+  public static final String PARAM_TYPE = "type";
+  public static final String PARAM_DESCRIPTION = "description";
+  public static final String PARAM_DOMAIN = "domain";
+
+  private static final String FIELD_ID = "id";
+  private static final String FIELD_NAME = PARAM_NAME;
```
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java]: Nicely done :+1:

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    try {
+      MetricDto metricInDb = dbClient.metricDao().selectNullableByKey(dbSession, key);
+      checkMetricInDb(metricInDb, key);
+
+      MetricDto metricTemplate = new MetricDto()
+        .setKey(key)
+        .setShortName(request.mandatoryParam(PARAM_NAME))
+        .setValueType(Metric.ValueType.valueOf(request.mandatoryParam(PARAM_TYPE)).name())
+        .setDescription(request.param(PARAM_DESCRIPTION))
+        .setDomain(request.param(PARAM_DOMAIN));
```
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java]: Indeed, domain is not marked as ``@Nullable``. Should it be mandatory?

File: server/sonar-server/src/test/java/org/sonar/server/metric/ws/CreateActionTest.java [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+    newRequest()
+      .setParam(PARAM_NAME, DEFAULT_NAME)
+      .setParam(PARAM_KEY, DEFAULT_KEY).execute();
+  }
+
+  @Test
+  public void fail_when_insufficient_privileges() throws Exception {
+    expectedException.expect(ForbiddenException.class);
+    userSessionRule.login("login").setGlobalPermissions(GlobalPermissions.SCAN_EXECUTION);
```
Comment (jblievremont) [inline server/sonar-server/src/test/java/org/sonar/server/metric/ws/CreateActionTest.java]: IMHO you can skip the ``setGlobalPermissions`` call: when you call ``login(String)``, the session is reset and all permissions are removed.

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    action.createParam(PARAM_KEY)
+      .setRequired(true)
+      .setDescription("Key")
+      .setExampleValue("team_size");
+
+    action.createParam(PARAM_TYPE)
+      .setRequired(true)
+      .setDescription("Metric type key")
+      .setPossibleValues(Metric.ValueType.names())
+      .setExampleValue(String.format("<ul><li>%s</li><li>%s</li></ul>", Metric.ValueType.INT.name(), Metric.ValueType.BOOL.name()));
```
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java]: Why a ``<ul/>`` here? Example value should be a single value.

**Comment by jblievremont** [comment]
Very minor feedback, then LGTM.

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java (hunk 2 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java]: thanks

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java (hunk 3 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java]: nope, clearer to read but I changed it to set when it's not null. I still find this rule cumbersome for dto. By default field are null so it makes sense to set directly. In the end the code is less readable but more explicit.

File: server/sonar-server/src/test/java/org/sonar/server/metric/ws/CreateActionTest.java (hunk 4 above)
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/metric/ws/CreateActionTest.java]: ok

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java (hunk 5 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java]: I wanted to display 2 different values, but ok just one

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    }
+  }
+
+  private MetricDto newMetricTemplate(Request request) {
+    MetricDto metricTemplate = new MetricDto()
+      .setKey(request.mandatoryParam(PARAM_KEY))
+      .setShortName(request.mandatoryParam(PARAM_NAME))
+      .setValueType(Metric.ValueType.valueOf(request.mandatoryParam(PARAM_TYPE)).name());
+    if ((request.param(PARAM_DOMAIN) != null)) {
+      metricTemplate.setDomain(request.param(PARAM_DOMAIN));
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java]: CRITICAL Possible null pointer dereference in org.sonar.server.metric.ws.CreateAction.newMetricTemplate(Request) due to return value of called method [rule](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE)

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+      MyBatis.closeQuietly(dbSession);
+    }
+  }
+
+  private MetricDto newMetricTemplate(Request request) {
+    MetricDto metricTemplate = new MetricDto()
+      .setKey(request.mandatoryParam(PARAM_KEY))
+      .setShortName(request.mandatoryParam(PARAM_NAME))
+      .setValueType(Metric.ValueType.valueOf(request.mandatoryParam(PARAM_TYPE)).name());
+    if ((request.param(PARAM_DOMAIN) != null)) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java]: MAJOR Remove those useless parentheses. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessParenthesesCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 1 critical
* 1 major

**Pull request merged**